### PR TITLE
Add newline support

### LINKED ISSUE #25: Terminal width option doesn't seem to work
```
$ python3
Python 3.2.3 (default, Feb 28 2014, 00:21:38)
[GCC 4.7.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from pyfiglet import Figlet; f = Figlet(font='standard', width=10); print (f.renderText('test'))
 _            _
| |_ ___  ___| |_
| __/ _ \/ __| __|
| ||  __/\__ \ |_
 \__\___||___/\__|



```
```
$ python
Python 2.7.3 (default, Feb 27 2014, 19:37:34)
[GCC 4.7.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from pyfiglet import Figlet; f = Figlet(font='standard', width=10); print f.renderText('test')
 _            _
| |_ ___  ___| |_
| __/ _ \/ __| __|
| ||  __/\__ \ |_
 \__\___||___/\__|



```
```
$ ./pyfiglet -w 10 -f standard test
 _            _
| |_ ___  ___| |_
| __/ _ \/ __| __|
| ||  __/\__ \ |_
 \__\___||___/\__|

```
```
# dpkg -l |grep figlet
ii  figlet                                         2.2.5-1                             amd64        Make large character ASCII banners out of ordinary text
# figlet -w 10 -f standard test
 _
| |_ ___
| __/ _ \
| ||  __/
 \__\___|

     _
 ___| |_
/ __| __|
\__ \ |_
|___/\__|

```

### PR DESCRIPTION
**Pull request by grandemk**
Added carriage return support to pyfiglet
refactored the FigletRenderingEngine: Use a Director - Builder - Product design pattern
refactored test.py, tests are now easy to add.
added a test_font script too diff fonts results with figlet
added stats_erroneous_font.py based on test.py, show which font and type of smush value are failing the tests 

If the width is smaller than a character width of the font, we throw an exception, get a bigger terminal with more than 6 columns :-) This is different from figlet which will either send many spaces or little part of the characters

Fixes #25.

### CODE REVIEW

**Comment by pwaller** [comment]
Wow, a lot happened here! It's a lot to digest. Sorry I haven't been fast to respond, I've been away from the computer. In the future I would recommend testing the water with smaller changes at a time, big changes are much riskier and so take much more time to process. If there is a problem with some part of it, it is harder to back that one part out. Bear in mind that this project has quite a number of uses in obscure places, so stability is important.

I'm happy in principle, but there are a few things I would like to address.

The first major one is that I am not keen to add extra dependencies. Could you please justify each dependency carefully, or alternatively just write it without them? I realise some of them are only needed for tests and whatnot, but in my mind, you should assume dependencies are expensive. They could break at some point in the future. (And this is a simple little project so pulling in external stuff massively increases the complexity!).

Secondly, would you be amenable to squashing your history a bit? There are lots of commits where I don't think so many are needed. There is also a great guideline to [making good commit messages here](http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html). I would recommend running `git rebase -i origin/master` and using the `fixup` and `reword` directives.

**Comment by pwaller** [comment]
Thanks again for looking into this :)

**Comment by grandemk** [comment]
I can erase all the dependencies as they were all here to ease testing (i
mean, i don't think matplotlib, colorama, etc. have something to do with
pyfiglet :) )
I think it is best to keep subprocess32 for python2 because it backports
many fixes from subprocess3 in python2. It's faster and it's recommended by
the python documentation: https://docs.python.org/2/library/subprocess.html,
but it's only for posix system so the import should fallback to subprocess
when that's not the case.

Each commit was tested for regression and should not introduce changes in
the way pyfiglet handle a correct one liner. The only difference is that
the width argument is now taken into consideration, so programs who used
this can break.

I agree that there are many changes, I initially tried to not change the
code structure too much but the result was horrible.
If I had to resume the changes, I would say that I put the content of a
while loop inside an object and that I replaced the buffer representing the
ascii art by a queue of buffer. I will try to make the git commit reflect
this.

Regards,
Kevin Grandemange

2015-04-06 19:55 GMT+02:00 Peter Waller <notifications@github.com>:

**Comment by grandemk** [comment]
I also had problems with utf-8 and pyfiglet, the main would fail with
characters like é or à instead of ignoring them, I will separate those
issues from the actual feature in the commits

2015-04-06 23:39 GMT+02:00 Kevin Grandemange <grandemange.kevin@gmail.com>:

**Comment by pwaller** [comment]
I don't see a strong argument for subprocess32, you say it's faster, but is it a bottleneck? If not, then it doesn't really matter.

Please ping the thread when I should take another look :)

**Comment by grandemk** [comment]
Depends on what you expect from the tests, I used them to know if I was
following the same behaviour as figlet on all fonts after each
modification, so they had to go as fast as possible, but in the context of
integration tests on a travis server, it doesn't really matters if it takes
one or two minutes longer.

2015-04-07 10:16 GMT+02:00 Peter Waller <notifications@github.com>:

**Comment by grandemk** [comment]
Will do that :)

2015-04-07 10:43 GMT+02:00 Kevin Grandemange <grandemange.kevin@gmail.com>:

**Comment by pwaller** [comment]
Does it take one or two minutes longer? That sounds fairly extreme. One possibility is to use subprocess32 if it's available, otherwise falling back to the builtin. But I'd prefer not to have it listed as a dependency anywhere.

**Comment by grandemk** [comment]
The one minute gap was when I tested it with random strings and with
threads and with optimizations in the handling of the output buffer of
subprocess against the older test suite. This was over a long period of
time (something like 30 mins if I remember correctly). This isn't a real
argument in the context of the tests here.

I always use subprocess32 when using python2.7 because it handles many
corner case better, case where std output is buggy or where pipes fail
(segfault, etc.). So this makes the testing code more robust. But this is
more useful when you're purpose is to crash a program not when you're
testing it's behavior against a norm.

I will suppress subprocess32 and see if this create some problems, I
remember putting it because of some strange things happening while getting
the output of pyfiglet, but maybe it was only me taking my precautions.
​

**Comment by pwaller** [comment]
@stefanor do you have any input on the packaging side before I accept such a large change? Could you give me any hints as to things I should check?

**Comment by grandemk** [comment]
squashed the history a little and I tried to keep the refactoring and the features in different commit
I made the color dependencies optionals and they are not used in the travis build. I deleted the subprocess32 dependencies as it doesn't add much in the context of these tests.

**Comment by pwaller** [comment]
It looks like you've added requirements.txt which is unnecessary?

**Comment by grandemk** [comment]
Hum, yeah I squashed two opposite commits on requirements.txt, so it must
be empty... That's the travis requirements (default travis will search for
this file and prompt a warning if it isn't there).
I was also wondering if you wanted to keep the diff script ?

**Comment by pwaller** [comment]
OK. I'll try and find time to do a last review when I can. If you don't hear from me in 3 days, ping me. Sorry I can't be faster right now!

**Comment by grandemk** [comment]
pinging you

2015-04-08 16:14 GMT+02:00 Peter Waller <notifications@github.com>:

File: pyfiglet/test.py [hunk 1]
```diff
@@ -56,27 +84,42 @@ def main():
...
+                           '(default: %default)')
+
+    opts, args = parser.parse_args()
+    test = Test(opts)
+    banner("TESTING one word")
+    test.check_text("foo")
+    banner("TESTING cut at space")
+    test.check_text("This is a very long text with many spaces and little words")
+    banner("TESTING cut at last char")
+    test.check_text("Averylongwordthatwillbecutatsomepoint I hope")
```
Comment (pwaller) [inline pyfiglet/test.py]: How about adding this test? :)

```python
    banner("TESTING newline")
    test.check_text("This text\nuses newlines")
```

Is it hard to fix?

**Comment by pwaller** [comment]
@grandemk thanks again for addressing my concerns. I'm happy to accept this, with one last detail: is it easy to make it work when you have newlines on the input?

```
$ pyfiglet $'This text\nuses newlines'
 _____ _     _       _            _                       
|_   _| |__ (_)___  | |_ _____  _| |_ _   _ ___  ___  ___ 
  | | | '_ \| / __| | __/ _ \ \/ / __| | | / __|/ _ \/ __|
  | | | | | | \__ \ | ||  __/>  <| |_| |_| \__ \  __/\__ \
  |_| |_| |_|_|___/  \__\___/_/\_\\__|\__,_|___/\___||___/
                                                          
                     _ _                 
 _ __   _____      _| (_)_ __   ___  ___ 
| '_ \ / _ \ \ /\ / / | | '_ \ / _ \/ __|
| | | |  __/\ V  V /| | | | | |  __/\__ \
|_| |_|\___| \_/\_/ |_|_|_| |_|\___||___/
                                         

$ figlet $'This text\nuses newlines'
 _____ _     _       _            _   
|_   _| |__ (_)___  | |_ _____  _| |_ 
  | | | '_ \| / __| | __/ _ \ \/ / __|
  | | | | | | \__ \ | ||  __/>  <| |_ 
  |_| |_| |_|_|___/  \__\___/_/\_\\__|
                                      
                                           _ _                 
 _   _ ___  ___  ___   _ __   _____      _| (_)_ __   ___  ___ 
| | | / __|/ _ \/ __| | '_ \ / _ \ \ /\ / / | | '_ \ / _ \/ __|
| |_| \__ \  __/\__ \ | | | |  __/\ V  V /| | | | | |  __/\__ \
 \__,_|___/\___||___/ |_| |_|\___| \_/\_/ |_|_|_| |_|\___||___/
```

**Comment by grandemk** [comment]
Yes, it's done, I added the test and fixed the -s option of the test script

2015-04-14 19:00 GMT+02:00 Peter Waller <notifications@github.com>:

**Comment by pwaller** [comment]
I just ran `pyfiglet/test.py` but there are a large number of failures since the last commit. Could you take a look and be sure to ping the thread when you want me to take another look.

**Comment by grandemk** [comment]
Hum, that's stange... I was rebasing the history just earlier, maybe that's
what caused it ?
I git cloned a fresh repository, all the tests are passing, so you
shouldn't have problems anymore

2015-04-14 19:56 GMT+02:00 Peter Waller <notifications@github.com>:

**Comment by pwaller** [comment]
Thanks for your contribution!

**Comment by pwaller** [comment]
I've cut a release, 0.7.3, and uploaded to pypi. Let me know if you spot any problems :)

**Comment by grandemk** [comment]
Thanks, I will let you know :)

2015-04-14 21:39 GMT+02:00 Peter Waller <notifications@github.com>:
